### ISSUE TITLE
Several not exported members from react-i18next

### ISSUE THREAD

**Issue by tpinne** [isaachinman/next-i18next#345]
## Describe the bug
I followed the guide lines in the readme to enable the lib in our typescript nextjs project. I'm getting 3 errors that `NamespaceExtractor`, `WithTranslationOptions` and `Subtract` are not exported in `react-i18next/src`. Imported by `next-i18next/types.d.ts`.

Looking [here](https://github.com/i18next/react-i18next/blob/v10.11.0/src/index.d.ts) I don't see any of those types exported by `react-i18next`.

The error messages I get:
```
[0] node_modules/next-i18next/types.d.ts(5,3): error TS2305: Module '"[...path...]/node_modules/react-i18next/src"' has no exported member 'NamespaceExtractor'.
[0] node_modules/next-i18next/types.d.ts(6,3): error TS2724: Module '"[...path...]/node_modules/react-i18next/src"' has no exported member 'WithTranslationOptions'. Did you mean 'UseTranslationOptions'?
[0] node_modules/next-i18next/types.d.ts(8,3): error TS2305: Module '"[...path...]/node_modules/react-i18next/src"' has no exported member 'Subtract'.
```

## Occurs in next-i18next version
next-i18next@0.42.0

## Steps to reproduce
```
yarn add next-i18next
```

Our **server.ts** according to readme.

```typescript
import * as express from 'express';
import * as next from 'next';
import { join } from 'path';
import { parse } from 'url';
// FROM README
import nextI18NextMiddleware from 'next-i18next/middleware';

import routes from './app/routes';
// FROM README
import nextI18next from './app/utilities/i18n';

const port = parseInt(process.env.PORT!, 10) || 3000;
const dev = process.env.NODE_ENV !== 'production';
const app = next({ dev, dir: 'app' });
const handle = routes.getRequestHandler(app);

const server = express();

const rootStaticFiles: string[] = [
  '/robots.txt',
  '/sitemap.xml',
  '/favicon.ico',
];

app.prepare().then(() => {
  // FROM README
  server.use(nextI18NextMiddleware(nextI18next));

  server.get('*', (req, res) => {
    const parsedUrl = parse(req.url, true);
    if (
      parsedUrl.pathname &&
      rootStaticFiles.indexOf(parsedUrl.pathname) > -1
    ) {
      const path = join(__dirname, 'app/static', parsedUrl.pathname);
      app.serveStatic(req, res, path);
    } else {
      handle(req, res);
    }
  });

  server.listen(port);
});
```

**./app/utilities/i18n** according to readme.

```typescript
import NextI18Next from 'next-i18next';

const NextI18NextInstance = new NextI18Next({
  defaultLanguage: 'de',
  otherLanguages: [],
});

export default NextI18NextInstance;

export const {
  appWithTranslation,
  withTranslation,
} = NextI18NextInstance;
```

## Expected behaviour
Importing types which actually exist in `react-i18next`.

**Comment by isaachinman** [comment]
Not sure how far down this rabbit hole we should go.

Are you willing to create a PR with your desired changes?

**Comment by tpinne** [comment]
It's the very first time I use your lib and react-i18next for that matter. So I have no knowledge at all of either library at the moment and maybe do more harm than good.

But currently `next-i18next` won't work with any TS project using it upgrading the lib to version `0.41.0` or higher.

I don't know, but is it neccessary to export `withTranslation` under `class NextI18Next` [here](https://github.com/isaachinman/next-i18next/blob/ed81b2f264d4cf49150613e0426899f16d21b1e0/types.d.ts#L46) ??? If not the problem should be solved, because all broken imports are used in there and it has the same signature as the orgiinal `withTranslation` from `react-i18next`.

Otherwise it should suffice to update your export of that function with the same signature as in the [original](https://github.com/i18next/react-i18next/blob/070624f4eb1ee96850f06498ffc3dcd1f109c7df/src/index.d.ts#L76). Where [this type](https://github.com/i18next/react-i18next/blob/070624f4eb1ee96850f06498ffc3dcd1f109c7df/src/index.d.ts#L4) needs to be recreated locally.

If you think either way would do no harm I can try to test it here locally and creating a PR.

**Comment by isaachinman** [comment]
> But currently next-i18next won't work with any TS project using it upgrading the lib to version 0.41.0 or higher.

Can you explain what you mean? The package will compile just fine.

**Comment by tpinne** [comment]
Using the lib in a TS project will include the `types.d.ts` automatically. And in that file are the mentioned imports which fail because they are not existent in `react-i18next`.

**Comment by isaachinman** [comment]
Yes, that much is clear. However this is not breaking and I can confirm that v0.41.0 and beyond will compile just fine in TypeScript consumers of `next-i18next`.

**Comment by tpinne** [comment]
Ok, but how is that possible?

There is `"react-i18next": "^10.9.0"` as dependency listed. This installs currently to `react-i18next@10.11.0`. 

And in that version are neither `NamespaceExtractor`, `WithTranslationOptions` nor `Subtract` existent in https://github.com/i18next/react-i18next/blob/v10.11.0/src/index.js or typed in https://github.com/i18next/react-i18next/blob/v10.11.0/src/index.d.ts .

If I modify the `types.d.ts` like in https://gist.github.com/tpinne/d6caa6aaae4bc5a028d4cf17ccc5e5ca it works because those non-existent imports are gone.

**Comment by isaachinman** [comment]
I cannot reproduce this. @tpinne Do you have a repository which I can debug?

**Comment by tpinne** [comment]
Ok, I will try to set a barebone example up

**Comment by tpinne** [comment]
* Clone https://github.com/tpinne/next-i18next-typescript
* yarn install
* npx tsc i18n.ts

In our case we have more complex version of the server.js in Typescript and import the file there and compile the `server.ts` before running `node server.js`. But as the imported `i18n.ts` gets transpiled the same way you can just compile the file directly to get the error messages.

**Comment by tpinne** [comment]
@isaachinman thanks for merging. Do you have an estimated ETA for the next release to use these changes through npm?

**Comment by isaachinman** [comment]
Released with v0.44.0.
